pyomo/contrib/alternative_solutions/solnpool.py

```diff
@@ -9,119 +9,1004 @@
+    best_value : float
+        Optional information to provide a starting best-discovered value for tolerance comparisons.
+        Defaults to a 'nan' value that the first added solution's value will replace.
+    """
+
+    def __init__(
+        self,
+        name=None,
+        as_solution=None,
+        counter=None,
+        *,
+        max_pool_size=None,
+        objective=None,
+        abs_tolerance=0.0,
+        rel_tolerance=None,
+        sense_is_min=True,
+        best_value=nan,
+    ):
+        super().__init__(name, as_solution, counter, policy="keep_best")
+        assert (max_pool_size is None) or (
+            max_pool_size >= 1
+        ), "max_pool_size must be None or positive integer"
+        self.max_pool_size = max_pool_size
+        self.objective = 0 if objective is None else objective
+        self.abs_tolerance = abs_tolerance
+        self.rel_tolerance = rel_tolerance
+        self.sense_is_min = sense_is_min
+        self.best_value = best_value
+        self._heap = []
+        self._unique_solutions = set()
+
+    @property
+    def pool_config(self):
+        return dict(
+            max_pool_size=self.max_pool_size,
+            objective=self.objective,
+            abs_tolerance=self.abs_tolerance,
+            rel_tolerance=self.rel_tolerance,
+            sense_is_min=self.sense_is_min,
+            best_value=self.best_value,
+        )
 
-    return solutions
+    def add(self, *args, **kwargs):
+        """
+        Add inputted solution to SolutionPool.
+        Relies on the instance as_solution conversion method to convert inputs to Solution Object.
+        If solution already present or outside tolerance of the best objective value, new solution is not added.
+        If input solution is new and within tolerance of the best objective value, the converted Solution object to the pool dictionary.
+        ID value for the solution generated as next increment of instance PoolCounter.
+        When pool size < max_pool_size, new solution is added without deleting old solutions.
+        When pool size == max_pool_size, new solution is added and oldest solution deleted.
+
+        Parameters
+        ----------
+        Input needs to match as_solution format from pool initialization.
+
+        Returns
+        ----------
+        None or int
+            None value corresponds to solution was already present and is ignored.
+            When not present, the ID value to match the added solution from the solution pool's PoolCounter.
+            The ID value is also the pool dictionary key for this solution.
+        """
+        soln = self._as_solution(*args, **kwargs)
+        #
+        # Return None if the solution has already been added to the pool
+        #
+        tuple_repn = soln._tuple_repn()
+        if tuple_repn in self._unique_solutions:
+            return None
+        self._unique_solutions.add(tuple_repn)
+        #
+        value = soln.objective(self.objective).value
```

Comment:
I'm confused: In the default case, this makes sense since `self.objective` is `0` and that in turn gets the 0th objective from the `ObjectiveInfo` object on the `Solution` class, so you get its value. But the docstring on this class says that `self.objective` could be a function, which doesn't seem like it would work here, or I'm missing something?